src/celeritas/field/RZPhiMapField.hh

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/field/RZPhiMapField.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cmath>
+
+#include "corecel/Constants.hh"
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/cont/EnumArray.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/grid/FindInterp.hh"
+#include "corecel/grid/NonuniformGrid.hh"
+#include "corecel/math/Algorithms.hh"
+#include "corecel/math/Quantity.hh"
+#include "corecel/math/Turn.hh"
+#include "celeritas/Types.hh"
+
+#include "RZPhiMapFieldData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Evaluate the value of magnetic field based on a volume-based RZPhi field
+ * map.
+ */
+class RZPhiMapField
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Real3 = Array<real_type, 3>;
+    using FieldParamsRef = NativeCRef<RZPhiMapFieldParamsData>;
+    //!@}
+
+  public:
+    // Construct with the shared map data
+    inline CELER_FUNCTION explicit RZPhiMapField(FieldParamsRef const& shared);
+
+    // Evaluate the magnetic field value for the given position
+    CELER_FUNCTION
+    inline Real3 operator()(Real3 const& pos) const;
+
+  private:
+    // Shared constant field map
+    FieldParamsRef const& params_;
+
+    NonuniformGrid<real_type> const grid_r_;
+    NonuniformGrid<real_type> const grid_z_;
+    NonuniformGrid<real_type> const grid_phi_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with the shared magnetic field map data.
+ */
+CELER_FUNCTION
+RZPhiMapField::RZPhiMapField(FieldParamsRef const& params)
+    : params_{params}
+    , grid_r_{params_.grids.axes[CylAxis::R], params_.grids.storage}
+    , grid_z_{params_.grids.axes[CylAxis::Z], params_.grids.storage}
+    , grid_phi_{params_.grids.axes[CylAxis::Phi], params_.grids.storage}
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the magnetic field vector for the given position.
+ *
+ * This does a 3-D interpolation on the input grid and reconstructs the
+ * magnetic field vector from the stored R, Z, and Phi components of the field.
+ * The result is in the native Celeritas unit system.
+ */
+CELER_FUNCTION auto RZPhiMapField::operator()(Real3 const& pos) const -> Real3
+{
+    CELER_ENSURE(params_);
+
+    Real3 value{0, 0, 0};
+
+    // Convert Cartesian to cylindrical coordinates
+    real_type r = hypot(pos[0], pos[1]);
+    // Ensure phi is in [0, 2\f$\pi\f$)
+    real_type phi = atan2(pos[1], pos[0]);
+    if (phi < 0)
+    {
+        phi += TwoPi::value();
+    }
```

Comment:
Not using `atan2Turn` because I am expecting a turn between [0;1] instead of [-0.5;0.5}

Reply:
You'd have to do the same thing right?
```suggestion
    auto phi = atan2turn(pos[1], pos[0]);
    if (phi < zero_quantity())
    {
        phi += Turn_t<real_type>{1};
    }
```
The whole point of using Turns is you don't have to accumulate floating point error by mucking with pi.

Reply:
Maybe it's actually better to use turn in [-0.5;0.5], that will be one less branch on GPU